### PR TITLE
Errors logs improvements

### PR DESCRIPTION
**Pull request by ilyapuchka**
Original PR #146 

Example error logs:

```
1:9: error: 'for' statements should use the following syntax 'for x in y where condition'.
Hello {% for name in %}{{ name }}, {% endfor %}!
         ^~~~~~~~~~~
```

```
invalid-base.html2:24: error: filter error
{% block body %}Body {{ target|unknown }} {% endblock %}
                        ^~~~~~~~~~~~~~

invalid-child-super.html2:25: error: filter error
{% block body %}Child {{ block.super }}{% endblock %}
                         ^~~~~~~~~~~
```

### CODE REVIEW

File: Sources/Context.swift [hunk 1]
```diff
@@ -3,8 +3,12 @@ public class Context {
   var dictionaries: [[String: Any?]]
 
   public let environment: Environment
+  
+  public var errorReporter: ErrorReporter {
+    return environment.errorReporter
+  }
 
-  init(dictionary: [String: Any]? = nil, environment: Environment? = nil) {
+  init(dictionary: [String: Any?]? = nil, environment: Environment? = nil) {
```
Comment (kylef) [inline Sources/Context.swift]: I've seen this change in multiple pull requests. Wanna split this out into a separate PR for discussion?

Comment (ilyapuchka) [reply Sources/Context.swift]: it's in #166 , I just based all other branches on its branch so that tests don't randomly fail, that's why you see them in each pr =)

**Comment by ilyapuchka** [comment]
@kylef I've simplified this a lot by storing filenames and token positions at parsing time instead of calculating them at rendering time, now it looks much closer to what you've started in #82 .

**Comment by ilyapuchka** [comment]
@kylef this started to fail on swift 3.1 with some creepy  compiler error. Maybe we should just drop it already?

**Comment by DivineDominion** [comment]
Whoa this looks cool. I'd like to +1 here so somebody from the team takes a look :)

**Comment by djbe** [comment]
@ilyapuchka Any idea how hard it'd be to rebase/rework this on master for you?
While this may not seem a "flashy" issue with fancy new features, it could end up being the most useful PR of all those that are open right now for template writers.

**Comment by ilyapuchka** [comment]
@djbe I've updated this with the latest master. If you are going to merge any PRs please merge this first.

File: Tests/StencilTests/ForNodeSpec.swift [hunk 2]
```diff
@@ -98,15 +98,15 @@ func testForNode() {
 
     $0.it("renders the given nodes while filtering items using where expression") {
         let nodes: [NodeType] = [VariableNode(variable: "item"), VariableNode(variable: "forloop.counter")]
-        let `where` = try parseExpression(components: ["item", ">", "1"], tokenParser: TokenParser(tokens: [], environment: Environment()))
+      let `where` = try parseExpression(components: ["item", ">", "1"], tokenParser: TokenParser(tokens: [], environment: Environment()), token: .text(value: "", at: .unknown))
```
Comment (djbe) [inline Tests/StencilTests/ForNodeSpec.swift]: Indentation?

File: Tests/StencilTests/ForNodeSpec.swift [hunk 3]
```diff
@@ -98,15 +98,15 @@ func testForNode() {
...
         let node = ForNode(resolvable: Variable("items"), loopVariables: ["item"], nodes: nodes, emptyNodes: [], where: `where`)
         try expect(try node.render(context)) == "2132"
     }
 
     $0.it("renders the given empty nodes when all items filtered out with where expression") {
         let nodes: [NodeType] = [VariableNode(variable: "item")]
         let emptyNodes: [NodeType] = [TextNode(text: "empty")]
-        let `where` = try parseExpression(components: ["item", "==", "0"], tokenParser: TokenParser(tokens: [], environment: Environment()))
+      let `where` = try parseExpression(components: ["item", "==", "0"], tokenParser: TokenParser(tokens: [], environment: Environment()), token: .text(value: "", at: .unknown))
```
Comment (djbe) [inline Tests/StencilTests/ForNodeSpec.swift]: Indentation?

File: Sources/ForTag.swift [hunk 4]
```diff
@@ -6,54 +6,56 @@ class ForNode : NodeType {
...
     }
 
-    let resolvable = try parser.compileResolvable(components[3])
-
-    let `where` = hasToken("where", at: 4)
-      ? try parseExpression(components: Array(components.suffix(from: 5)), tokenParser: parser)
-      : nil
-
-    return ForNode(resolvable: resolvable, loopVariables: loopVariables, nodes: forNodes, emptyNodes:emptyNodes, where: `where`)
+    return ForNode(resolvable: resolvable, loopVariables: loopVariables, nodes: forNodes, emptyNodes:emptyNodes, where: `where`, token: token)
```
Comment (djbe) [inline Sources/ForTag.swift]: Space after the `:`.

File: Sources/ForTag.swift [hunk 5]
```diff
@@ -6,54 +6,56 @@ class ForNode : NodeType {
...
   }
 
-  init(resolvable: Resolvable, loopVariables: [String], nodes:[NodeType], emptyNodes:[NodeType], where: Expression? = nil) {
+  init(resolvable: Resolvable, loopVariables: [String], nodes:[NodeType], emptyNodes:[NodeType], where: Expression? = nil, token: Token? = nil) {
```
Comment (djbe) [inline Sources/ForTag.swift]: Space after the `:`.

File: Sources/Lexer.swift [hunk 6]
```diff
@@ -156,4 +179,24 @@ extension String {
...
+    var lineNumber: UInt = 0
+    var offset: Int = 0
+    var lineContent = ""
+
+    for line in components(separatedBy: CharacterSet.newlines) {
+      lineNumber += 1
+      lineContent = line
+      if let rangeOfLine = self.range(of: line), rangeOfLine.contains(range.lowerBound) {
+        offset = distance(from: rangeOfLine.lowerBound, to:
+          range.lowerBound)
```
Comment (djbe) [inline Sources/Lexer.swift]: bizarre spacing here?

File: Sources/Node.swift [hunk 7]
```diff
@@ -1,35 +1,31 @@
...
+
+  /// Reference to this node's token
+  var token: Token? { get }
 }
 
 
 /// Render the collection of nodes in the given context
 public func renderNodes(_ nodes:[NodeType], _ context:Context) throws -> String {
-  return try nodes.map { try $0.render(context) }.joined(separator: "")
+  return try nodes.map({
```
Comment (djbe) [inline Sources/Node.swift]: Those `()` aren't needed for the `map`.

File: Sources/Parser.swift [hunk 8]
```diff
@@ -110,15 +116,42 @@ public class TokenParser {
     return filtersWithDistance.filter({ $0.distance == minDistance }).map({ $0.filterName })
   }
 
+  public func compileFilter(_ filterToken: String, containedIn containingToken: Token) throws -> Resolvable {
+    do {
+      return try FilterExpression(token: filterToken, parser: self)
+    } catch {
+      if var error = error as? TemplateSyntaxError, error.token == nil {
```
Comment (djbe) [inline Sources/Parser.swift]: Might be good to use a guard here? It avoids 1 level of indentation.

File: Tests/StencilTests/EnvironmentSpec.swift [hunk 9]
```diff
@@ -32,9 +39,304 @@ func testEnvironment() {
 
       try expect(result) == "here"
     }
+
+    func expectedSyntaxError(token: String, template: Template, description: String) -> TemplateSyntaxError {
+      let range = template.templateString.range(of: token)!
```
Comment (djbe) [inline Tests/StencilTests/EnvironmentSpec.swift]: Always avoid implicit unwraps. You can replace this with a guard that falls back to an error with less information if the token isn't found. Or a `fatalError` (as this is a test, might as well break early).

File: Tests/StencilTests/EnvironmentSpec.swift [hunk 10]
```diff
@@ -32,9 +39,304 @@ func testEnvironment() {
...
+      let sourceMap = SourceMap(filename: template.name, line: rangeLine)
+      let token = Token.block(value: token, at: sourceMap)
+      return TemplateSyntaxError(reason: description, token: token, stackTrace: [])
+    }
+
+    func expectError(reason: String, token: String,
+                     file: String = #file, line: Int = #line, function: String = #function) throws {
+      let expectedError = expectedSyntaxError(token: token, template: template, description: reason)
+      
+      let error = try expect(environment.render(template: template, context: ["names": ["Bob", "Alice"], "name": "Bob"]),
```
Comment (djbe) [inline Tests/StencilTests/EnvironmentSpec.swift]: Both the declaration of this `func` as well as this line have some bizarre indentation. This also applies to the similar functions below.

Comment (djbe) [reply Tests/StencilTests/EnvironmentSpec.swift]: This function is repeated quite a few times. That could be rewritten?

File: Tests/StencilTests/LexerSpec.swift [hunk 11]
```diff
@@ -66,25 +69,26 @@ func testLexer() {
...
-        try expect(tokens[2]) == Token.variable(value: "name")
-        try expect(tokens[3]) == Token.block(value: "endif")
-        try expect(tokens[4]) == Token.text(value: ".")
+          "    if name\n" +
+          "     and\n" +
+          "    name\n" +
+          "%}{{\n" +
+          "name\n" +
+          "}}{%\n" +
+        "endif %}."
```
Comment (djbe) [inline Tests/StencilTests/LexerSpec.swift]: Indentation for this last line.

**Comment by djbe** [comment]
Seeing as you'd rather merge this first, I'll review this one before the others.

File: Tests/StencilTests/EnvironmentSpec.swift (hunk 10 above)
Comment (ilyapuchka) [reply Tests/StencilTests/EnvironmentSpec.swift]: I'd prefer this to 6 arguments stacked on each other
this can be cleared up sometime later, this PR is big without it

File: Tests/StencilTests/EnvironmentSpec.swift (hunk 9 above)
Comment (ilyapuchka) [reply Tests/StencilTests/EnvironmentSpec.swift]: I would prefer failing here as it is crystal clear what is the reason than pollute the code with boilerplate guards especially considering that this is a test and we don't need graceful error handling or whatever.

**Comment by ilyapuchka** [comment]
@djbe fixed indentations

File: Sources/Environment.swift [hunk 12]
```diff
@@ -3,9 +3,15 @@ public struct Environment {
   public var extensions: [Extension]
 
   public var loader: Loader?
+  public var errorReporter: ErrorReporter
+
+  public init(loader: Loader? = nil,
+              extensions: [Extension]? = nil,
+              templateClass: Template.Type = Template.self,
+              errorReporter: ErrorReporter = SimpleErrorReporter()) {
```
Comment (kylef) [inline Sources/Environment.swift]: If I understand correctly, the errorReporter property is only used by context's errorReporter property. But that isn't used anywhere (I may have missed this)?

What is the reason that the Environment and Context should hold the errorReporter? If I understand how this is intended to be used, a consumer of the library would catch an error and pass it to the errorReporter themselves? In that case I think it would make sense to remove the property from Environment and Context leaving the consumer free to create their own ErrorReporter as desired. Decoupling these a little bit.

**Comment by ilyapuchka** [comment]
@kylef this is probably a leftover from when I was using it to render error in inheritance nodes. Not needed now, removed in 92ebfe5

File: Tests/StencilTests/EnvironmentSpec.swift (hunk 9 above)
Comment (djbe) [reply Tests/StencilTests/EnvironmentSpec.swift]: A `guard` with a `fatalError` and a message then. It's not because it's a test that it shouldn't be good code.

Comment (ilyapuchka) [reply Tests/StencilTests/EnvironmentSpec.swift]: I think this code is good enough, but I will do that if that's what it takes to merge this.

Comment (ilyapuchka) [reply Tests/StencilTests/EnvironmentSpec.swift]: @djbe done in 96a004e

**Comment by ilyapuchka** [comment]
@kylef you'd like to add something?

**Comment by djbe** [comment]
Seeing as there's no more feedback, and this PR is GTM for me, we can go on to the next PRs 🎉
